### ISSUE TITLE
Ancillary Variables - validate 'all other' cube operations

### ISSUE THREAD

**Issue by trexfeathers** [SciTools/iris#3554]
**Status update 2019-12-09**
It appears that *most* of the points raised + discussed below may now have been ruled out.

**Current 'TODO' list (including things already-complete) :**
  * [x] ~'other' indexing-like cube methods : `Cube.subset` and `Cube.intersection`~
    * **believed ok, nothing to do** (see below)
  * [x] ~regrid/interpolate : Cube.regid, Cube.interpolate and all schemes code~
    * **believed ok, nothing to do** (see below)
  * [x] `Cube.xml()` : #3607
  * [x] possible uses of `Cube.slices` / `slices_over` : **TBD**
    * two places still of concern,
      * ~[this line](https://github.com/SciTools/iris/blob/479b2cf80fc75c7b327c01a0f6a2b139ea6e9896/lib/iris/analysis/_regrid.py#L477) uses `cube.slices()` and result is re-merged later.  Waiting for #3569 to assess potential failure of the split-and-remerge~
        * **believed ok, nothing to do** ([this line](https://github.com/SciTools/iris/blob/479b2cf80fc75c7b327c01a0f6a2b139ea6e9896/lib/iris/analysis/_regrid.py#L483) removes ancillary variables)
      * [this line](https://github.com/SciTools/iris/blob/479b2cf80fc75c7b327c01a0f6a2b139ea6e9896/lib/iris/analysis/stats.py#L120) uses `cube.slices_over()`: **will be fixed by #3621**
        * was ~Waiting on #3478 to see possible effect of this on result cube~, but #3621 means we don't need to wait for that.
  * [x] user documentation of behaviours : #3608
    * note: #3570 has broader scope, but I think we can agree that if there is more here, it is not MVP.

**Notes:**
  * This ticket applies to "all" cube operations that produce a new cube, where an input cube with ancillaries might give a result with "wrong" ancillaries.  For instance, where "old" ancillaries could now have the wrong dimensions.
  * It also covers some documentation, somehow, of "what" operations will do with ancils.
    * However, we may want to defer those additions until we are ready to "release" ancils, as described below.  See [separate ticket #3570](https://github.com/SciTools/iris/issues/3570)
    * It _might_ also be desirable to add matching notes for cell-measures handling, which are currently absent pretty much anywhere.  

**Use of [launch_ancils feature-branch](https://github.com/SciTools/iris/pull/3556#issue-343470988)**
Pending a final decision to "release" ancils in Iris 3.0, we need to defer documentation changes to the feature branch.  See key note [here](https://github.com/SciTools/iris/pull/3556#issue-343470988) 
That applies also to the documentation of cube operations, as mentioned below : see [separate ticket #3570 to cover this](https://github.com/SciTools/iris/issues/3570)

---
**Original notes** :
**original name : "Ancillary Variables - cube.copy invalid retention of AV's #3554"**

The following include uses of `cube.copy`, which in #3546 will newly retain Cell Measures and Ancillary Variables. In these cases this could be inappropriate since the CM's/AV's could refer to dimensions that no longer exist.

As noted in #3483 there is a wider discussion about regridding/interpolation to be had since depending on the dimensions altered it may be appropriate to retain CM's/AV's. Until that is tackled in detail it is important the below operations do not behave in a misleading way and therefore a blanket discarding of CM's/AV's would be safest.

- `analysis._regrid.CurvilinearRegridder`
- `analysis.trajectory.UnstructuredNearestNeigbourRegridder`, also `__call__` method
- `experimental.stratify.relevel`
<p>&nbsp;</p>

> - [x] to be started after #3546 is merged.  confirm whether this list has other cube.copy dependencies are used in interpolation.
> - [x] check whether uses of the above return cubes with cell measures or ancillary variables
>
> If the above **are** returning cubes with CM's/AV's:
> - [ ] alter the functions/methods to prevent retention of CM's/AV's
> - [ ] write tests to demonstrate this
> - [ ] document this behavior
> - [ ] ensure that the cases covered here are accounted for in any further discussions/work to conditionally retain/discard CM's/AV's depending on the dimensions involved

**Comment by stephenworsley** [comment]
From what I can tell looking at the code, (I haven't run tests yet) `analysis._regrid.CurvilinearRegridder` should discard any ancillary variables unless code is added to this function to account for them:
https://github.com/SciTools/iris/blob/2d75d7f979ae992155286fdc20307bfb0b71ff02/lib/iris/analysis/_regrid.py#L290 The last time the copy is used is in this code which makes a fresh cube using only the shape of the copy (which is here called `grid_cube`): https://github.com/SciTools/iris/blob/2d75d7f979ae992155286fdc20307bfb0b71ff02/lib/iris/analysis/_regrid.py#L355-L358

**Comment by stephenworsley** [comment]
`experimental.stratify.relevel` also looks like it shouldn't copy over ancillary variables since the only time a cube is copied, only its metadata is extracted. https://github.com/SciTools/iris/blob/2d75d7f979ae992155286fdc20307bfb0b71ff02/lib/iris/experimental/stratify.py#L175

**Comment by stephenworsley** [comment]
When `analysis.trajectory.UnstructuredNearestNeigbourRegridder` is called, the copied source cube is first passed to `interpolate`: https://github.com/SciTools/iris/blob/2d75d7f979ae992155286fdc20307bfb0b71ff02/lib/iris/analysis/trajectory.py#L873-L875 The resulting cube is then recreated from its data, metadata and coordinates: https://github.com/SciTools/iris/blob/2d75d7f979ae992155286fdc20307bfb0b71ff02/lib/iris/analysis/trajectory.py#L883-L898
This `result_cube` is then returned. It shouldn't contain any ancillary variables. The only problems left would have to be inside the `interpolate` call.

**Comment by pp-mo** [comment]
> @stephenworsley "experimental.stratify.relevel also looks like it shouldn't copy over ancillary variables since the only time a cube is copied, only its metadata is extracted. "
> `result = Cube(new_data, **cube.copy().metadata._asdict())`

That frankly looks like a bit of bug to me + a stupid thing to do : copy the whole cube to get an indepenedent (scalar) metadata copy, ? really ?
The only reason I can see for copying the cube before taking metadata it is that the _as_dict() does not copy objects in the dict values, which I guess would only really affect cell_methods anyway. It would make sense to fix that with a deepcopy of the dict, I would think.
A better coding solution should really be found. Especially as `_as_dict` isn't even a public method.

**Comment by stephenworsley** [comment]
One more thing to be aware of, `iris.util.new_axis` is currently discarding cell measures and ancillary variables. This is probably worth addressing if we're expecting people to use it and concatenate to bypass the restrictions around merge.

**Comment by stephenworsley** [comment]
One more thing that's probably worth noting. The `iris.cube.Cube` methods `remove_ancillary_variable`, `ancillary_variable_dims` and `cell_measure_dims` do not currently accept a name (string) as a valid argument. See #3295 for a simmilar case where this was already adressed for `remove_cell_measure`.

**Comment by pp-mo** [comment]
Ticked all the boxes, at last !
